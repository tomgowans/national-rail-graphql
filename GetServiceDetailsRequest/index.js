const fetchData = require("../fetchData");
const removeArrays = require("../removeArrays");

/**
 *
 * @param {string} serviceID (string): The LDBWS service ID of the service to request the details of. The service ID is obtained from a service listed in a StationBoard object returned from any other request.
 */

function GetServiceDetailsRequest({ serviceID }, tokenValue) {
  return fetchData({
    body: `
      <ldb:GetServiceDetailsRequest>
        ${serviceID && `<ldb:serviceID>${serviceID}</ldb:serviceID>`}
      </ldb:GetServiceDetailsRequest>
      `,
    tokenValue
  })
    .then(result => {
      // console.dir(result);
      return removeArrays(
        result.GetServiceDetailsResponse[0].GetServiceDetailsResult[0]
      );
    })
    .catch(err => console.error(err));
}

module.exports = GetServiceDetailsRequest;
