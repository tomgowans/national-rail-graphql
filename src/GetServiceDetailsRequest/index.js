const fetchData = require("../fetchData");
const removeArrays = require("../removeArrays");

/**
 *
 * @param {string} serviceID (string): The LDBWS service ID of the service to request the details of. The service ID is obtained from a service listed in a StationBoard object returned from any other request.
 */

function GetServiceDetailsRequest({ serviceID }, tokenValue) {
  return new Promise((resolve, reject) => {
    fetchData({
      body: `
      <ldb:GetServiceDetailsRequest>
        ${serviceID && `<ldb:serviceID>${serviceID}</ldb:serviceID>`}
      </ldb:GetServiceDetailsRequest>
      `,
      tokenValue,
    })
      .then((result) => {
        // console.dir(result);
        resolve(
          removeArrays(
            result.GetServiceDetailsResponse[0].GetServiceDetailsResult[0]
          )
        );
      })
      .catch((err) => reject(err));
  });
}

module.exports = GetServiceDetailsRequest;
