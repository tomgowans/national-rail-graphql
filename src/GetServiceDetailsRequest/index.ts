import fetchData from '../fetchData/index';
import removeArrays from '../removeArrays';
import { Result, ServiceAttributes } from '../types';

/**
 *
 * @param {string} serviceID (string): The LDBWS service ID of the service to request the details of. The service ID is obtained from a service listed in a StationBoard object returned from any other request.
 */

function GetServiceDetailsRequest({ serviceID }: ServiceAttributes, tokenValue: string): Promise<Result> {
  return new Promise((resolve, reject) => {
    fetchData({
      body: `
      <ldb:GetServiceDetailsRequest>
        ${serviceID && `<ldb:serviceID>${serviceID}</ldb:serviceID>`}
      </ldb:GetServiceDetailsRequest>
      `,
      tokenValue,
    })
      .then((result: Result) => {
        // console.dir(result);
        resolve(removeArrays(result.GetServiceDetailsResponse[0].GetServiceDetailsResult[0]));
      })
      .catch((err: Error) => reject(err));
  });
}

export default GetServiceDetailsRequest;
