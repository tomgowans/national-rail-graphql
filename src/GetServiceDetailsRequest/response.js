const response = {
  text: `<?xml version="1.0" encoding="utf-8"?>
  <soap:Envelope xmlns:soap="http://www.w3.org/2003/05/soap-envelope" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:xsd="http://www.w3.org/2001/XMLSchema">
      <soap:Body>
          <GetServiceDetailsResponse xmlns="http://thalesgroup.com/RTTI/2017-10-01/ldb/">
              <GetServiceDetailsResult xmlns:lt="http://thalesgroup.com/RTTI/2012-01-13/ldb/types" xmlns:lt6="http://thalesgroup.com/RTTI/2017-02-02/ldb/types" xmlns:lt7="http://thalesgroup.com/RTTI/2017-10-01/ldb/types" xmlns:lt4="http://thalesgroup.com/RTTI/2015-11-27/ldb/types" xmlns:lt5="http://thalesgroup.com/RTTI/2016-02-16/ldb/types" xmlns:lt2="http://thalesgroup.com/RTTI/2014-02-20/ldb/types" xmlns:lt3="http://thalesgroup.com/RTTI/2015-05-14/ldb/types">
                  <lt7:generatedAt>2020-02-27T08:31:54.4790478+00:00</lt7:generatedAt>
                  <lt7:serviceType>train</lt7:serviceType>
                  <lt7:locationName>East Croydon</lt7:locationName>
                  <lt7:crs>ECR</lt7:crs>
                  <lt7:operator>Southern</lt7:operator>
                  <lt7:operatorCode>SN</lt7:operatorCode>
                  <lt7:length>10</lt7:length>
                  <lt7:platform>6</lt7:platform>
                  <lt7:sta>09:03</lt7:sta>
                  <lt7:eta>On time</lt7:eta>
                  <lt7:std>09:04</lt7:std>
                  <lt7:etd>On time</lt7:etd>
                  <lt7:previousCallingPoints>
                      <lt7:callingPointList>
                          <lt7:callingPoint>
                              <lt7:locationName>London Bridge</lt7:locationName>
                              <lt7:crs>LBG</lt7:crs>
                              <lt7:st>08:47</lt7:st>
                              <lt7:et>On time</lt7:et>
                              <lt7:length>10</lt7:length>
                          </lt7:callingPoint>
                      </lt7:callingPointList>
                  </lt7:previousCallingPoints>
                  <lt7:subsequentCallingPoints>
                      <lt7:callingPointList>
                          <lt7:callingPoint>
                              <lt7:locationName>South Croydon</lt7:locationName>
                              <lt7:crs>SCY</lt7:crs>
                              <lt7:st>09:06</lt7:st>
                              <lt7:et>On time</lt7:et>
                              <lt7:length>10</lt7:length>
                          </lt7:callingPoint>
                          <lt7:callingPoint>
                              <lt7:locationName>Purley Oaks</lt7:locationName>
                              <lt7:crs>PUO</lt7:crs>
                              <lt7:st>09:09</lt7:st>
                              <lt7:et>On time</lt7:et>
                              <lt7:length>10</lt7:length>
                          </lt7:callingPoint>
                          <lt7:callingPoint>
                              <lt7:locationName>Purley</lt7:locationName>
                              <lt7:crs>PUR</lt7:crs>
                              <lt7:st>09:12</lt7:st>
                              <lt7:et>On time</lt7:et>
                              <lt7:length>5</lt7:length>
                          </lt7:callingPoint>
                          <lt7:callingPoint>
                              <lt7:locationName>Kenley</lt7:locationName>
                              <lt7:crs>KLY</lt7:crs>
                              <lt7:st>09:18</lt7:st>
                              <lt7:et>On time</lt7:et>
                              <lt7:length>5</lt7:length>
                          </lt7:callingPoint>
                          <lt7:callingPoint>
                              <lt7:locationName>Whyteleafe</lt7:locationName>
                              <lt7:crs>WHY</lt7:crs>
                              <lt7:st>09:22</lt7:st>
                              <lt7:et>On time</lt7:et>
                              <lt7:length>5</lt7:length>
                          </lt7:callingPoint>
                          <lt7:callingPoint>
                              <lt7:locationName>Whyteleafe South</lt7:locationName>
                              <lt7:crs>WHS</lt7:crs>
                              <lt7:st>09:24</lt7:st>
                              <lt7:et>On time</lt7:et>
                              <lt7:length>5</lt7:length>
                          </lt7:callingPoint>
                          <lt7:callingPoint>
                              <lt7:locationName>Caterham</lt7:locationName>
                              <lt7:crs>CAT</lt7:crs>
                              <lt7:st>09:28</lt7:st>
                              <lt7:et>On time</lt7:et>
                              <lt7:length>5</lt7:length>
                          </lt7:callingPoint>
                      </lt7:callingPointList>
                      <lt7:callingPointList>
                          <lt7:callingPoint>
                              <lt7:locationName>Purley</lt7:locationName>
                              <lt7:crs>PUR</lt7:crs>
                              <lt7:st>09:12</lt7:st>
                              <lt7:et>On time</lt7:et>
                              <lt7:length>5</lt7:length>
                          </lt7:callingPoint>
                          <lt7:callingPoint>
                              <lt7:locationName>Reedham (Surrey)</lt7:locationName>
                              <lt7:crs>RHM</lt7:crs>
                              <lt7:st>09:20</lt7:st>
                              <lt7:et>On time</lt7:et>
                              <lt7:length>5</lt7:length>
                          </lt7:callingPoint>
                          <lt7:callingPoint>
                              <lt7:locationName>Coulsdon Town</lt7:locationName>
                              <lt7:crs>CDN</lt7:crs>
                              <lt7:st>09:23</lt7:st>
                              <lt7:et>On time</lt7:et>
                              <lt7:length>5</lt7:length>
                          </lt7:callingPoint>
                          <lt7:callingPoint>
                              <lt7:locationName>Woodmansterne</lt7:locationName>
                              <lt7:crs>WME</lt7:crs>
                              <lt7:st>09:26</lt7:st>
                              <lt7:et>On time</lt7:et>
                              <lt7:length>5</lt7:length>
                          </lt7:callingPoint>
                          <lt7:callingPoint>
                              <lt7:locationName>Chipstead</lt7:locationName>
                              <lt7:crs>CHP</lt7:crs>
                              <lt7:st>09:29</lt7:st>
                              <lt7:et>On time</lt7:et>
                              <lt7:length>5</lt7:length>
                          </lt7:callingPoint>
                          <lt7:callingPoint>
                              <lt7:locationName>Kingswood</lt7:locationName>
                              <lt7:crs>KND</lt7:crs>
                              <lt7:st>09:35</lt7:st>
                              <lt7:et>On time</lt7:et>
                              <lt7:length>5</lt7:length>
                          </lt7:callingPoint>
                          <lt7:callingPoint>
                              <lt7:locationName>Tadworth</lt7:locationName>
                              <lt7:crs>TAD</lt7:crs>
                              <lt7:st>09:38</lt7:st>
                              <lt7:et>On time</lt7:et>
                              <lt7:length>5</lt7:length>
                          </lt7:callingPoint>
                          <lt7:callingPoint>
                              <lt7:locationName>Tattenham Corner</lt7:locationName>
                              <lt7:crs>TAT</lt7:crs>
                              <lt7:st>09:42</lt7:st>
                              <lt7:et>On time</lt7:et>
                              <lt7:length>5</lt7:length>
                          </lt7:callingPoint>
                      </lt7:callingPointList>
                  </lt7:subsequentCallingPoints>
              </GetServiceDetailsResult>
          </GetServiceDetailsResponse>
      </soap:Body>
  </soap:Envelope>`,
};

module.exports = response;
