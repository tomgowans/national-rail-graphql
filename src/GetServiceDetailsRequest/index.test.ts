import GetServiceDetailsRequest from "./index";

describe("GetServiceDetailsRequest", () => {
  it("sends a request with minimal parameters correctly", async () => {
    expect.assertions(1);

    const result = await GetServiceDetailsRequest(
      {
        serviceID: "ABC123"
      },
      "TOKEN_VALUE"
    );
    expect(result).toMatchSnapshot();
  });

  it("sends a request with minimal parameters and fail", async () => {
    expect.assertions(1);

    await expect(
      GetServiceDetailsRequest(
        {
          serviceID: "ABC123"
        },
        ""
      )
    ).rejects.toThrow("Error with credentials");
  });

  it("sends a request with more parameters correctly", () => {
    expect.assertions(1);

    return expect(
      GetServiceDetailsRequest(
        {
          serviceID: "ABC123"
        },
        "TOKEN_VALUE"
      )
    ).resolves.toMatchSnapshot();
  });
});
