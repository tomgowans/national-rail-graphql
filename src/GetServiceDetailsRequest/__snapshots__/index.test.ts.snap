// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GetServiceDetailsRequest sends a request with minimal parameters correctly 1`] = `
Object {
  "$": Object {
    "xmlns:lt": "http://thalesgroup.com/RTTI/2012-01-13/ldb/types",
    "xmlns:lt2": "http://thalesgroup.com/RTTI/2014-02-20/ldb/types",
    "xmlns:lt3": "http://thalesgroup.com/RTTI/2015-05-14/ldb/types",
    "xmlns:lt4": "http://thalesgroup.com/RTTI/2015-11-27/ldb/types",
    "xmlns:lt5": "http://thalesgroup.com/RTTI/2016-02-16/ldb/types",
    "xmlns:lt6": "http://thalesgroup.com/RTTI/2017-02-02/ldb/types",
    "xmlns:lt7": "http://thalesgroup.com/RTTI/2017-10-01/ldb/types",
  },
  "crs": "ECR",
  "eta": "On time",
  "etd": "On time",
  "generatedAt": "2020-02-27T08:31:54.4790478+00:00",
  "length": 10,
  "locationName": "East Croydon",
  "operator": "Southern",
  "operatorCode": "SN",
  "platform": 6,
  "previousCallingPoints": Object {
    "callingPointList": Array [
      Object {
        "callingPoint": Array [
          Object {
            "crs": "LBG",
            "et": "On time",
            "length": 10,
            "locationName": "London Bridge",
            "st": "08:47",
          },
        ],
      },
    ],
  },
  "serviceType": "train",
  "sta": "09:03",
  "std": "09:04",
  "subsequentCallingPoints": Object {
    "callingPointList": Array [
      Object {
        "callingPoint": Array [
          Object {
            "crs": "SCY",
            "et": "On time",
            "length": 10,
            "locationName": "South Croydon",
            "st": "09:06",
          },
          Object {
            "crs": "PUO",
            "et": "On time",
            "length": 10,
            "locationName": "Purley Oaks",
            "st": "09:09",
          },
          Object {
            "crs": "PUR",
            "et": "On time",
            "length": 5,
            "locationName": "Purley",
            "st": "09:12",
          },
          Object {
            "crs": "KLY",
            "et": "On time",
            "length": 5,
            "locationName": "Kenley",
            "st": "09:18",
          },
          Object {
            "crs": "WHY",
            "et": "On time",
            "length": 5,
            "locationName": "Whyteleafe",
            "st": "09:22",
          },
          Object {
            "crs": "WHS",
            "et": "On time",
            "length": 5,
            "locationName": "Whyteleafe South",
            "st": "09:24",
          },
          Object {
            "crs": "CAT",
            "et": "On time",
            "length": 5,
            "locationName": "Caterham",
            "st": "09:28",
          },
        ],
      },
      Object {
        "callingPoint": Array [
          Object {
            "crs": "PUR",
            "et": "On time",
            "length": 5,
            "locationName": "Purley",
            "st": "09:12",
          },
          Object {
            "crs": "RHM",
            "et": "On time",
            "length": 5,
            "locationName": "Reedham (Surrey)",
            "st": "09:20",
          },
          Object {
            "crs": "CDN",
            "et": "On time",
            "length": 5,
            "locationName": "Coulsdon Town",
            "st": "09:23",
          },
          Object {
            "crs": "WME",
            "et": "On time",
            "length": 5,
            "locationName": "Woodmansterne",
            "st": "09:26",
          },
          Object {
            "crs": "CHP",
            "et": "On time",
            "length": 5,
            "locationName": "Chipstead",
            "st": "09:29",
          },
          Object {
            "crs": "KND",
            "et": "On time",
            "length": 5,
            "locationName": "Kingswood",
            "st": "09:35",
          },
          Object {
            "crs": "TAD",
            "et": "On time",
            "length": 5,
            "locationName": "Tadworth",
            "st": "09:38",
          },
          Object {
            "crs": "TAT",
            "et": "On time",
            "length": 5,
            "locationName": "Tattenham Corner",
            "st": "09:42",
          },
        ],
      },
    ],
  },
}
`;

exports[`GetServiceDetailsRequest sends a request with more parameters correctly 1`] = `
Object {
  "$": Object {
    "xmlns:lt": "http://thalesgroup.com/RTTI/2012-01-13/ldb/types",
    "xmlns:lt2": "http://thalesgroup.com/RTTI/2014-02-20/ldb/types",
    "xmlns:lt3": "http://thalesgroup.com/RTTI/2015-05-14/ldb/types",
    "xmlns:lt4": "http://thalesgroup.com/RTTI/2015-11-27/ldb/types",
    "xmlns:lt5": "http://thalesgroup.com/RTTI/2016-02-16/ldb/types",
    "xmlns:lt6": "http://thalesgroup.com/RTTI/2017-02-02/ldb/types",
    "xmlns:lt7": "http://thalesgroup.com/RTTI/2017-10-01/ldb/types",
  },
  "crs": "ECR",
  "eta": "On time",
  "etd": "On time",
  "generatedAt": "2020-02-27T08:31:54.4790478+00:00",
  "length": 10,
  "locationName": "East Croydon",
  "operator": "Southern",
  "operatorCode": "SN",
  "platform": 6,
  "previousCallingPoints": Object {
    "callingPointList": Array [
      Object {
        "callingPoint": Array [
          Object {
            "crs": "LBG",
            "et": "On time",
            "length": 10,
            "locationName": "London Bridge",
            "st": "08:47",
          },
        ],
      },
    ],
  },
  "serviceType": "train",
  "sta": "09:03",
  "std": "09:04",
  "subsequentCallingPoints": Object {
    "callingPointList": Array [
      Object {
        "callingPoint": Array [
          Object {
            "crs": "SCY",
            "et": "On time",
            "length": 10,
            "locationName": "South Croydon",
            "st": "09:06",
          },
          Object {
            "crs": "PUO",
            "et": "On time",
            "length": 10,
            "locationName": "Purley Oaks",
            "st": "09:09",
          },
          Object {
            "crs": "PUR",
            "et": "On time",
            "length": 5,
            "locationName": "Purley",
            "st": "09:12",
          },
          Object {
            "crs": "KLY",
            "et": "On time",
            "length": 5,
            "locationName": "Kenley",
            "st": "09:18",
          },
          Object {
            "crs": "WHY",
            "et": "On time",
            "length": 5,
            "locationName": "Whyteleafe",
            "st": "09:22",
          },
          Object {
            "crs": "WHS",
            "et": "On time",
            "length": 5,
            "locationName": "Whyteleafe South",
            "st": "09:24",
          },
          Object {
            "crs": "CAT",
            "et": "On time",
            "length": 5,
            "locationName": "Caterham",
            "st": "09:28",
          },
        ],
      },
      Object {
        "callingPoint": Array [
          Object {
            "crs": "PUR",
            "et": "On time",
            "length": 5,
            "locationName": "Purley",
            "st": "09:12",
          },
          Object {
            "crs": "RHM",
            "et": "On time",
            "length": 5,
            "locationName": "Reedham (Surrey)",
            "st": "09:20",
          },
          Object {
            "crs": "CDN",
            "et": "On time",
            "length": 5,
            "locationName": "Coulsdon Town",
            "st": "09:23",
          },
          Object {
            "crs": "WME",
            "et": "On time",
            "length": 5,
            "locationName": "Woodmansterne",
            "st": "09:26",
          },
          Object {
            "crs": "CHP",
            "et": "On time",
            "length": 5,
            "locationName": "Chipstead",
            "st": "09:29",
          },
          Object {
            "crs": "KND",
            "et": "On time",
            "length": 5,
            "locationName": "Kingswood",
            "st": "09:35",
          },
          Object {
            "crs": "TAD",
            "et": "On time",
            "length": 5,
            "locationName": "Tadworth",
            "st": "09:38",
          },
          Object {
            "crs": "TAT",
            "et": "On time",
            "length": 5,
            "locationName": "Tattenham Corner",
            "st": "09:42",
          },
        ],
      },
    ],
  },
}
`;
